Rework the pipeline tutorial (#20437)
huggingface/transformers: docs/source/en/pipeline_tutorial.mdx

```py
>>> from transformers import pipeline

>>> generator = pipeline(task="automatic-speech-recognition")
```

2. Pass your input text to the [`pipeline`]:

```py
>>> generator("https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/mlk.flac")
{'text': 'I HAVE A DREAM BUT ONE DAY THIS NATION WILL RISE UP LIVE UP THE TRUE MEANING OF ITS TREES'}
```

Not perfect for what you had in mind ? Check out the models on the hub [Most downloaded models for automatic speech recognition](https://huggingface.co/models?pipeline_tag=automatic-speech-recognition&sort=downloads). 

> Let's take [openai/whisper-large](https://huggingface.co/openai/whisper-large) to try it out.

Author: I like the emphasis on "why" we want to change. It's because the output of the audio is incorrect towards the end.

Also the models are downloaded and it is a sort of ranking which is important. I don't care about model X downloaded 3 times and that handles nothing and was just a test experiment for X. I want to widely used common place models.

I feel those two informations are important.

> Let's take [openai/whisper-large](https://huggingface.co/openai/whisper-large) to try it out.

Reviewer: For sure, that makes sense!

I updated my suggestion a bit to keep those two important points. Mainly, I like making it super obvious to readers that we are talking about the result here and why they should check the most downloaded models. Feel free to ignore if you prefer it as is though 🙂

```py
>>> generator = pipeline(model="openai/whisper-large")
>>> generator("https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/mlk.flac")
{'text': ' I have a dream that one day this nation will rise up and live out the true meaning of its creed.'}
```

Now this is more appropriate !
Really check out the hub, for different languages, models specialized in your field and more.
And you can check out the live results directly from the hub to see if it fits or 
handle corner cases better than another one.
If you don't find any, start [`training`]

If you have several inputs, you can pass your input as a list:

```py
generator(
    [
        "https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/mlk.flac",
        "https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/1.flac",
    ]
)
```

> If you want to iterate over a whole dataset, or want to use it for inference in a webserver, check out dedicated parts

Author: Again, putting link in their own lines, make them more readable, more scannable by lazy readers. 

I think I prefer it the current way. 

I basically never click links within text.

```py
generator(model="openai/whisper-large", device=0)
```

If the model is too large for a single GPU, you can also use [`accelerate`](https://huggingface.co/docs/accelerate/index) (`device_map="auto"`).

```py
#!pip install accelerate
generator(model="openai/whisper-large", device_map="auto")
```

### Batch size

By default, pipelines will not batch inference for reasons explained in detail [here](https://huggingface.co/docs/transformers/main_classes/pipelines#pipeline-batching). The reason is that batching is not necessarily faster, and can actually be quite slower in some cases.

But if it works in your use case, you can use:

```py
generator(model="openai/whisper-large", device=0, batch_size=2)
audio_filenames = [f"audio_{i}.flac" for i in range(10)]
texts = generator(audio_filenames)
```

This will actually run the pipeline on all 10 audio files you are providing, but it will pass them by batches of 2
to the model (which is on GPU where batching is more likely to help), without any further code from you. 
The output should always match what you would have received without batching. It is only meant as a way to help 
get more speed out of the pipeline. 

> Pipelines alleviate some complexity of batching, because for some pipelines, 
> a single item (like a long audio file) need to be chunked into multiple parts
> for processing by a model. Pipeline [abstract that away for you](./main_classes/pipelines#pipeline-chunk-batching)

Reviewer: ```suggestion
Pipelines can also alleviate some of the complexities of batching because, for some pipelines, a single item (like a long audio file) needs to be chunked into multiple parts to be processed by a model. The pipeline performs this [*chunk batching*](./main_classes/pipelines#pipeline-chunk-batching) for you.
```

### Task specific parameters

> All tasks provide task specific parameters which should help you get the job done.

Reviewer: ```suggestion
All tasks provide task specific parameters which allow for additional flexibility and options to help you get your job done.
```

> For instance, the [`transformers.AutomaticSpeechRecognitionPipeline.__call__`] method
> 
> Here we can see `return_timestamps` which sounds promising for subtitling videos.

Reviewer: ```suggestion
For instance, the [`transformers.AutomaticSpeechRecognitionPipeline.__call__`] method has a `return_timestamps` parameter which sounds promising for subtitling videos:
```

{'text': 'I HAVE A DREAM BUT ONE DAY THIS NATION WILL RISE UP AND LIVE OUT THE TRUE MEANING OF ITS CREED',
 'chunks': [
      {'text': 'I', 'timestamp': (1.22, 1.24)},
      {'text': 'HAVE', 'timestamp': (1.42, 1.58)},
      {'text': 'A', 'timestamp': (1.66, 1.68)},
      {'text': 'DREAM', 'timestamp': (1.76, 2.14)},
      {'text': 'BUT', 'timestamp': (3.68, 3.8)},
      {'text': 'ONE', 'timestamp': (3.94, 4.06)},
      {'text': 'DAY', 'timestamp': (4.16, 4.3)},
      {'text': 'THIS', 'timestamp': (6.36, 6.54)},
      {'text': 'NATION', 'timestamp': (6.68, 7.1)},
      {'text': 'WILL', 'timestamp': (7.32, 7.56)},
      {'text': 'RISE', 'timestamp': (7.8, 8.26)},
      {'text': 'UP', 'timestamp': (8.38, 8.48)},
      {'text': 'AND', 'timestamp': (10.08, 10.18)},
      {'text': 'LIVE', 'timestamp': (10.26, 10.48)},
      {'text': 'OUT', 'timestamp': (10.58, 10.7)},
      {'text': 'THE', 'timestamp': (10.82, 10.9)},
      {'text': 'TRUE', 'timestamp': (10.98, 11.18)},
      {'text': 'MEANING', 'timestamp': (11.26, 11.58)},
      {'text': 'OF', 'timestamp': (11.66, 11.7)},
      {'text': 'ITS', 'timestamp': (11.76, 11.88)},
      {'text': 'CREED', 'timestamp': (12.0, 12.38)}
]}
```

> With that parameter, the model inferred the texts, but also outputted along **when** the various words were pronounced

Reviewer: ```suggestion
As you can see, the model inferred the text and also outputted **when** the various words were pronounced
```

in the sentence.

> There are many parameters for all tasks, really check out the docs to find them.
> For instance this pipeline also handles `chunk_length_s` which is helpful to work on 
> really long audio files which the models cannot handle on their own. This is helpful when
> subtitling entire movies or hour-long videos.

Reviewer: ```suggestion
There are many parameters available for each task, so check out each task's API reference to see what you can tinker with!
For instance, the [`~transformers.AutomaticSpeechRecognitionPipeline`] has a `chunk_length_s` parameter which is helpful for working on really long audio files (for example, subtitling entire movies or hour-long videos) that a model typically cannot handle on its own.
```

> Cannot you find a parameter which would really help you out ?
> [Request it](https://github.com/huggingface/transformers/issues/new?assignees=&labels=feature&template=feature-request.yml)

Reviewer: ```suggestion
If you can't find a parameter that would really help you out, feel free to [request it](https://github.com/huggingface/transformers/issues/new?assignees=&labels=feature&template=feature-request.yml)!
```

## Using pipelines on a dataset

> So you have a large dataset and want to run inference on it to get something like a
> metric for it, perfect !
> 
> The easiest and more generic way we recomment is using an iterator.

Reviewer: ```suggestion
The pipeline can also run inference on a large dataset. The easiest way we recommend doing this is by using an iterator:
```

```py
def data():
    for i in range(1000):
        yield f"My example {i}"

pipe = pipe(model="gpt2", device=0)
generated_characters = 0
for out in pipe(data()):
    generated_characters += len(out["generated_text"])
```

> The iterator `data()` will yield each result. The pipeline will automatically
> recognize that the input is iterable, and will start fetching the data while
> it continues to process it on the GPU. This uses [DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) under the hood.
> This is important so you don't have to allocate memory for the whole dataset
> and you can feed the GPU as fast as possible.

Reviewer: ```suggestion
The iterator `data()` yields each result, and the pipeline automatically
recognizes the input is iterable and will start fetching the data while
it continues to process it on the GPU (this uses [DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) under the hood).
This is important because you don't have to allocate memory for the whole dataset
and you can feed the GPU as fast as possible.
```

> Since batching could speed things up, using `batch_size` here and tuning it could be important.

Reviewer: ```suggestion
Since batching could speed things up, it may be useful to try tuning the `batch_size` parameter here.
```

> Turns out we also have a fantasic library for [datasets](https://github.com/huggingface/datasets/).
> which makes this even simpler.

Reviewer: ```suggestion
The simplest way to iterate over a dataset is to just load one from 🤗 [Datasets](https://github.com/huggingface/datasets/):
```

```py
# KeyDataset is a util that will just output the item we're interested in.
from transformers.pipelines.pt_utils import KeyDataset

pipe = pipeline(model="hf-internal-testing/tiny-random-wav2vec2", device=0)
dataset = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation[:10]")

for out in pipe(KeyDataset(dataset["audio"])):
    print(out)
```

## Using pipelines for a webserver

<Tip>
Creating an inference engine is a complex topic which deserves it's own
page.
</Tip>

[Link](./pipeline_webserver)

> ### Vision pipeline

Reviewer: This section and the ones below can be ##-level headers since they don't really fit under the topic above.

```suggestion
## Vision pipeline
```

Reviewer: Oops I think you forgot to change the section level for the Text and Multimodal sections!

```py
>>> from transformers import pipeline

> >>> generator = pipeline(task="automatic-speech-generation")

Reviewer: ```suggestion
>>> generator = pipeline(task="automatic-speech-recognition")
```

```

2. Pass your input text to the [`pipeline`]:

```py
>>> generator(

> ...     "https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/mlk.flac"

Reviewer 2: Is there any way to have a widget/link to a widget where the user could listen to the sample?

Author: Well if you use this link it should play directly in your browser. 
Does so for FF at least. (It's  a hub configuration away if it doesn't)

```

Now this is more appropriate !
Really check out the hub, for different languages, models specialized in your field and more.
And you can check out the live results directly from the hub to see if it fits or 
handle corner cases better than another one.
If you don't find any, start [`fine-tuning`]  # TODO FIX link.

If you have several inputs, you can pass your input as a list:

```py
>>> generator(
...     [
...         "https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/mlk.flac",
...         "https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/1.flac"]
...     ]
... )  # doctest: +SKIP
```

If you want to iterate over a whole dataset, or want to use it for inference in a webserver, check out dedicated parts

[`pipeline_dataset`]
[`pipeline_webserver`]

## Parameters

[`pipeline`] supports many parameters, some are task specific, some are general for all pipelines.
In general you can specify parameters anywhere you want:

```py
>>> generator(model="openai/whisper-large", my_parameter=1)
>>> out = generate(...) # This will use `my_parameter=1`.
>>> out = generate(..., my_parameter=2) # This will override and use `my_parameter=2`.

> >>> out = generate(...) # This will go back to using `my_parameter=1`.

Reviewer 2: Remove the doctest syntax here as this is not executable.

Author: Done.

```

Let's check out 3 important ones:

### Device

If you use `device=n` the pipeline will automatically put the model on said device.
This will work regardless wether you are using `PyTorch` or `Tensorflow`.

> If the model is too large for a single GPU, you can also use [`accelerate`] (`device_map="auto"`).

Reviewer: Super cool, didn't know this was possible! I'd include a code sample here as well because it's not really clear if you have to use `device_map` with Accelerate or if you can just use it within the `pipeline`

# TODO check link.

```py

> >>> generator(model="openai/whisper-large", device=0)

Reviewer 2: Same here as this will only work when a GPU is available.

Author: Done.

Reviewer 2: Looks like you forgot to push ;-)

```

### Batch size

By default, pipelines will not batch inference for reasons explained in detail [here](https://huggingface.co/docs/transformers/main_classes/pipelines#pipeline-batching). The reason is that batching is not necessarily faster, and can actually be quite slower in some cases.

But if it works in your use case, you can use:

```py

> >>> generator(model="openai/whisper-large", device=0, batch_size=2)

Reviewer 2: Same here, unless the doctests are run on a machine with GPU @ydshieh ?

Reviewer 3: doctest CI is run on a GPU machine. But it might be easier to make it work on CPU too - so the team members can debug more easily whenever something have to be fixed.

(sometimes it's easier/quicker to access a CPU machine - although I think everyone have GPU machines)

Author: I still removed it, not sure there's a lot of value added  in checking those values (it's tested against in the CI by fast tests anyway for all pipelines)

>>> audio_filenames = [f"audio_{i}.flac" for i in range(10)]
>>> texts = generator(audio_filenames)
```

This will actually run the pipeline on all 10 audio files you are providing, but it will pass them by batches of 2
to the model (which is on GPU where batching is more likely to help), without any further code from you. 
The output should always match what you would have received without batching. It is only meant as a way to help 
get more speed out of the pipeline. 

> ### Task specific `return_timestamps`.

Reviewer: ```suggestion
### Task specific parameters
```

All tasks provide task specific parameters which should help you get the job done.

> For instance [https://huggingface.co/docs/transformers/main_classes/pipelines#transformers.AutomaticSpeechRecognitionPipeline.__call__]() # TODO fix link

Reviewer: I gotchu :)

```suggestion
For instance, the [`transformers.AutomaticSpeechRecognitionPipeline.__call__`] method
```

pipe = pipe(model="gpt2", device=0)
generated_characters = 0
for out in pipe(data()):
    generated_characters += len(out["generated_text"])
```

The iterator `data()` will yield each result. The pipeline will automatically
recognize that the input is iterable, and will start fetching the data while
it continues to process it on the GPU. This uses [DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) under the hood.
This is important so you don't have to allocate memory for the whole dataset
and you can feed the GPU as fast as possible.

Since batching could speed things up, using `batch_size` here and tuning it could be important.

Turns out we also have a fantasic library for [datasets](https://github.com/huggingface/datasets/).
which makes this even simpler.

```py
# KeyDataste is a util that will just output the item we're interested in.
from transformers.pipelines.pt_utils import KeyDataset

pipe = pipeline(model="hf-internal-testing/tiny-random-wav2vec2", device=0)
dataset = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation[:10]")

for out in pipe(KeyDataset(dataset["audio"])):
   print(out)
```

> ## Using pipelines for a webserver

Reviewer 2: This is too advanced for this tutorial. It can go in another page however as this is certainly of interest for some users!

[{'score': 0.4335, 'label': 'lynx, catamount'}, {'score': 0.0348, 'label': 'cougar, puma, catamount, mountain lion, painter, panther, Felis concolor'}, {'score': 0.0324, 'label': 'snow leopard, ounce, Panthera uncia'}, {'score': 0.0239, 'label': 'Egyptian cat'}, {'score': 0.0229, 'label': 'tiger cat'}]
```

### Text pipeline

> Using a [`pipeline`] for nlp tasks is practically identical.

Reviewer: ```suggestion
Using a [`pipeline`] for NLP tasks is practically identical.
```

```py
>>> from transformers import pipeline

>>> # This model is a `zero-shot-classification` model.
>>> # It will classify text, except you are free to choose any label you might imagine
>>> classifier = pipeline(model="facebook/bart-large-mnli")
>>> classifier(
...     "I have a problem with my iphone that needs to be resolved asap!!",
...     candidate_labels=["urgent", "not urgent", "phone", "tablet", "computer"],
... )
{'sequence': 'I have a problem with my iphone that needs to be resolved asap!!',
 'labels': ['urgent', 'phone', 'computer', 'not urgent', 'tablet'],
 'scores': [0.504,0.479,0.013,0.003,0.002]}
```

### Multimodal pipeline

The [`pipeline`] supports more than one modality. For example, a visual question answering (VQA) task combines text and image. Feel free to use any image link you like and a question you want to ask about the image. The image can be a URL or a local path to the image.

> For example, if you use the same image from the vision pipeline above:

Reviewer: ```suggestion
For example, if you use this [invoice image](https://huggingface.co/spaces/impira/docquery/resolve/2359223c1837a7587402bda0f2643382a6eefeab/invoice.png):
```

[{'score': 0.4335, 'label': 'lynx, catamount'}, {'score': 0.0348, 'label': 'cougar, puma, catamount, mountain lion, painter, panther, Felis concolor'}, {'score': 0.0324, 'label': 'snow leopard, ounce, Panthera uncia'}, {'score': 0.0239, 'label': 'Egyptian cat'}, {'score': 0.0229, 'label': 'tiger cat'}]
```

> ### Text pipeline

Author: Moved this to last, since we are still too much "NLP" for my taste.
